riscv/src/runtime.rs

```diff
@@ -73,635 +68,20 @@ impl SubMachine {
-
-        let handlers = self.syscalls.iter().flat_map(|(syscall, implementation)| {
-            std::iter::once(format!("__ecall_handler_{syscall}:"))
-                .chain(implementation.0.iter().map(|i| i.to_string()))
-                .chain([format!("jump_dyn 1, {};", Register::from("tmp1").addr())])
-        });
-
-        ecall
-            .chain(jump_table)
-            .chain(invalid_handler)
-            .chain(handlers)
-            .chain(std::iter::once("// end of ecall handler".to_string()))
-            .collect()
-    }
-
-    pub fn submachine_names(&self) -> String {
-        self.submachines.keys().join("\n")
-    }
-}
-
-impl TryFrom<&[&str]> for Runtime {
-    type Error = String;
-
-    fn try_from(names: &[&str]) -> Result<Self, Self::Error> {
-        let mut runtime = Runtime::base();
-        for name in names {
-            if runtime.has_submachine(name) {
-                continue;
-            }
-            match *name {
-                "poseidon_gl" => runtime = runtime.with_poseidon_no_continuations(),
-                "keccakf" => runtime = runtime.with_keccak(),
-                "arith" => runtime = runtime.with_arith(),
-                _ => return Err(format!("Invalid co-processor specified: {name}")),
-            }
-        }
-        Ok(runtime)
-    }
-}
-
-/// Helper function for register names used in instruction params
-fn reg(idx: usize) -> String {
-    format!("{EXTRA_REG_PREFIX}{idx}")
-}
-
-/// Helper function to generate instr link for large number input/output registers
-fn instr_link(call: &str, inputs: usize, outputs: usize) -> String {
-    format!(
-        "{}{}({})",
-        if outputs > 0 {
-            format!(
-                "({}) = ",
-                (0..outputs).map(|i| format!("{}'", reg(i))).join(", ")
-            )
-        } else {
-            "".to_string()
-        },
-        call,
-        (0..inputs).map(reg).join(", ")
-    )
-}
-
-/// Load word from addr+offset into register
-fn load_word(addr_reg_id: u32, offset: u32, reg: &str) -> [String; 2] {
-    let tmp1 = Register::from("tmp1");
-    let tmp2 = Register::from("tmp2");
-    [
-        format!(
-            "mload {addr_reg_id}, {offset}, {}, {};",
-            tmp1.addr(),
-            tmp2.addr()
-        ),
-        format!("{reg} <=X= get_reg({});", tmp1.addr()),
-    ]
-}
-
-/// Store word from register into addr+offset
-fn store_word(addr_reg_id: u32, offset: u32, reg: &str) -> [String; 3] {
-    let tmp1 = Register::from("tmp1");
-    let tmp2 = Register::from("tmp2");
-    [
-        // split_gl ensures we store a 32-bit value
-        format!("set_reg {}, {reg};", tmp1.addr()),
-        format!(
-            "split_gl {}, {}, {};",
-            tmp1.addr(),
-            tmp1.addr(),
-            tmp2.addr()
-        ),
-        format!("mstore {addr_reg_id}, 0, {offset}, {};", tmp1.addr()),
-    ]
+#[derive(Clone)]
+pub struct SyscallImpl(pub Vec<FunctionStatement>);
+
+pub trait Runtime {
```

Comment:
This trait could even go away potentially, since we don't use it as dyn or Box anymore.

Reply:
removed now to not overengineer the abstraction